test/agent/cmd/snat-utils/main.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-iptables/iptables"
+)
+
+func main() {
+	var testIPTableRules bool
+	var testExternalDomainConnectivity bool
+	var randomizedSNATValue string
+	var url string
+
+	flag.BoolVar(&testIPTableRules, "testIPTableRules", false, "bool flag when set to true tests validate if IPTable has required rules")
+	flag.StringVar(&randomizedSNATValue, "randomizedSNATValue", "prng", "value for AWS_VPC_K8S_CNI_RANDOMIZESNAT")
+	flag.BoolVar(&testExternalDomainConnectivity, "testExternalDomainConnectivity", false, "bool flag when set to true tests if the pod has internet access")
+	flag.StringVar(&url, "url", "https://aws.amazon.com/", "url to check for connectivity")
+
+	flag.Parse()
+
+	if testIPTableRules {
+		err := validateIPTableRules(randomizedSNATValue)
+		if err != nil {
+			log.Fatal(err)
+		}
+		log.Printf("Randomized SNAT test passed for AWS_VPC_K8S_CNI_RANDOMIZESNAT: %s\n", randomizedSNATValue)
+	}
+
+	if testExternalDomainConnectivity {
+		err := validateExternalDomainConnectivity(url)
+		if err != nil {
+			log.Fatal(err)
+		}
+		log.Println("External Domain Connectivity test passed")
+	}
+}
+
+func validateExternalDomainConnectivity(url string) error {
+	timeout := time.Duration(120 * time.Second)
+	client := http.Client{
+		Timeout: timeout,
+	}
+	resp, err := client.Get(url)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("%s returned response code: %d", url, resp.StatusCode)
+	}
+	return nil
+}
+
+func validateIPTableRules(randomizedSNATValue string) error {
+	// Check IPTable rules corresponding to AWS_VPC_K8S_CNI_RANDOMIZESNAT
+	expectedString := "random-fully"
+	iptables, err := iptables.New()
+	if err != nil {
+		return err
+	}
+
+	if !iptables.HasRandomFully() || randomizedSNATValue == "hashrandom" {
+		expectedString = "random"
+	}
+
+	chains, err := iptables.List("nat", "AWS-SNAT-CHAIN-1")
+	if err != nil {
+		return err
+	}
+
+	containsExpectedString := false
+	rule := ""
+	for _, chain := range chains {
+		if strings.Contains(chain, expectedString) {
+			rule = chain
+			containsExpectedString = true
+			break
+		}
+	}
+
+	if randomizedSNATValue == "none" && containsExpectedString {
+		return fmt.Errorf("failed: found unexpected %s for SNAT rule: %s", expectedString, rule)
+	} else if randomizedSNATValue != "none" && !containsExpectedString {
+		return fmt.Errorf("failed: did not find expected %s for any of the SNAT rules", expectedString)
+	}
+	return nil
```

Comment:
We should construct the expected IP table rules (chains) and validate for the entire rule set. We appear to be only checking for the presence of `random-fully` and not the entire rule?

https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/networkutils/network.go#L348

Reply:
Checking